runnable/extra-files/dexcept.h

```diff
@@ -0,0 +1,63 @@
+// D exceptions to by caught by C++
+
+#pragma once
+
+namespace D {
+
+struct string
+{
+    size_t length;
+    const char* ptr;
+};
+
+class Object;
+
+extern "C" {
+    string _d_toString(Object* o);
+}
+
+class Object
+{
+    void* __monitor;
+
+    virtual void _classinfo_data() = 0;
+
+protected:
+    // don't call directly, ABI mismatch

```

Comment:
If I understand https://github.com/rainers/druntime/commit/9dce356c4132310b01147b2941205f4c55b602eb correctly, it allows you to catch a D exception in C++ by declaring a C++ type with the same class name as the native D exception in a `D` namespace (convention © Rainer ;)). That's why each thrown exception's type incl. base types are 'mangled' twice.
The definition of the C++ type doesn't really matter, as long as the exception object isn't accessed; Rainer's test only uses `toString()`, for which he wrote a specific C++ implementation anyway (calling back into druntime, apparently due to ABI mismatches). By duplicating the D class definitions of `Object`, `Throwable` etc. I guess he wanted to pave the way for a catch-D-exceptions lib for C++.